File: android/app/src/main/java/com/dailysns/services/DataService.kt
package com.dailysns.services

import com.dailysns.models.*
import kotlinx.coroutines.delay

/**
 * DataService.kt
 * DailySNSWithYourSS Android版
 * 
 * データサービス（モック実装）
 */

interface DataServiceProtocol {
    suspend fun fetchAlbums(): List<Album>
    suspend fun fetchRandomFollowingPosts(userId: String): List<Post>
    suspend fun fetchRandomGlobalPosts(): List<Post>
    suspend fun createPost(post: Post): Post
    suspend fun fetchUser(id: String): User
}

class DataService : DataServiceProtocol {
    
    /**
     * アルバムを取得（モック）
     */
    override suspend fun fetchAlbums(): List<Album> {
        delay(500)
        return mockAlbums()
    }
    
    /**
     * フォロー中のユーザーの投稿を取得（モック）
     */
    override suspend fun fetchRandomFollowingPosts(userId: String): List<Post> {
        delay(500)
        val allPosts = mockPosts()
        return allPosts.take(minOf(5, allPosts.size))
    }
    
    /**
     * グローバル投稿を取得（モック）
     */
    override suspend fun fetchRandomGlobalPosts(): List<Post> {
        delay(500)
        return mockPosts()
    }
    
    /**
     * 投稿を作成（モック）
     */
    override suspend fun createPost(post: Post): Post {
        delay(500)
        return post
    }
    
    /**
     * ユーザーを取得（モック）
     */
    override suspend fun fetchUser(id: String): User {
        delay(500)
        return mockUsers().find { it.id == id }
            ?: User(username = "unknown", displayName = "Unknown User")
    }
    
    // モックデータ生成
    
    private fun mockUsers(): List<User> {
        return listOf(
            User(id = "1", username = "user1", displayName = "Alice"),
            User(id = "2", username = "user2", displayName = "Bob"),
            User(id = "3", username = "user3", displayName = "Charlie")
        )
    }
    
    private fun mockPosts(): List<Post> {
        val users = mockUsers()
        val posts = mutableListOf<Post>()
        
        users.forEach { user ->
            repeat(3) { i ->
                val items = listOf(
                    PostItem(type = PostItemType.SCREENSHOT, content = "screenshot_$i.jpg"),
                    PostItem(type = PostItemType.COMMENT, content = "This is a comment $i")
                )
                posts.add(Post(userId = user.id, items = items))
            }
        }
        
        return posts
    }
    
    private fun mockAlbums(): List<Album> {
        val users = mockUsers()
        return users.map { user ->
            val userPosts = mockPosts().filter { it.userId == user.id }
            Album(
                userId = user.id,
                userName = user.displayName,
                userProfileImageURL = user.profileImageURL,
                posts = userPosts
            )
        }
    }
}

File: web/src/services/DataService.ts
/**
 * DataService.ts
 * DailySNSWithYourSS Web版
 * 
 * データサービス（モック実装）
 */

import { Post, createPost, createPostItem, PostItemType } from '../models/Post';
import { User, Album, createUser, createAlbum } from '../models/User';

export interface DataServiceProtocol {
  fetchAlbums(): Promise<Album[]>;
  fetchRandomFollowingPosts(userId: string): Promise<Post[]>;
  fetchRandomGlobalPosts(): Promise<Post[]>;
  createPost(post: Post): Promise<Post>;
  fetchUser(id: string): Promise<User>;
}

export class DataService implements DataServiceProtocol {
  /**
   * アルバムを取得（モック）
   */
  async fetchAlbums(): Promise<Album[]> {
    await this.delay(500);
    return this.mockAlbums();
  }

  /**
   * フォロー中のユーザーの投稿を取得（モック）
   */
  async fetchRandomFollowingPosts(userId: string): Promise<Post[]> {
    await this.delay(500);
    const allPosts = this.mockPosts();
    return allPosts.slice(0, Math.min(5, allPosts.length));
  }

  /**
   * グローバル投稿を取得（モック）
   */
  async fetchRandomGlobalPosts(): Promise<Post[]> {
    await this.delay(500);
    return this.mockPosts();
  }

  /**
   * 投稿を作成（モック）
   */
  async createPost(post: Post): Promise<Post> {
    await this.delay(500);
    return post;
  }

  /**
   * ユーザーを取得（モック）
   */
  async fetchUser(id: string): Promise<User> {
    await this.delay(500);
    const users = this.mockUsers();
    return users.find(u => u.id === id) || createUser('unknown', 'Unknown User');
  }

  // モックデータ生成

  private delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  private mockUsers(): User[] {
    return [
      createUser('user1', 'Alice', '1'),
      createUser('user2', 'Bob', '2'),
      createUser('user3', 'Charlie', '3')
    ];
  }

  private mockPosts(): Post[] {
    const users = this.mockUsers();
    const posts: Post[] = [];

    users.forEach(user => {
      for (let i = 0; i < 3; i++) {
        const items = [
          createPostItem(PostItemType.Screenshot, `screenshot_${i}.jpg`),
          createPostItem(PostItemType.Comment, `This is a comment ${i}`)
        ];
        posts.push(createPost(user.id, items));
      }
    });

    return posts;
  }

  private mockAlbums(): Album[] {
    const users = this.mockUsers();
    return users.map(user => {
      const userPosts = this.mockPosts().filter(p => p.userId === user.id);
      return createAlbum(
        user.id,
        user.displayName,
        userPosts,
        undefined,
        user.profileImageURL
      );
    });
  }
}
